src/mem/cache/prefetch/ipcp.cc

```diff
@@ -0,0 +1,398 @@
+                }
+            }
+        }
+        //update region of current PC
+        if (ipcpTable[ipIndex].regionAddr != regionBase) {
+            ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+             || RST[regionBase].trained);
+            ipcpTable[ipIndex].regionAddr = regionBase;
+        }
+
+    } else {
+        RST[regionBase].lastTouch = curTick();
+        RST[regionBase].cacheLineTouched.insert(blkAddr);
+        RST[regionBase].dir = 32;
+        RST[regionBase].trained = false;
+        RST[regionBase].lastAddr = blkAddr;
+
+        //Visit IP Table and check if previous was trained?
+        if (RST.find(ipcpTable[ipIndex].regionAddr) != RST.end()) {
+            RST[regionBase].tentative =
+             RST[ipcpTable[ipIndex].regionAddr].trained;
+        }
+
+        ipcpTable[ipIndex].regionAddr = regionBase;
+        ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+         || RST[regionBase].trained);
+     }
+
+    //Check for victim
+    if (RST.size() > RM_SIZE) {
+        //dummy victim
+        std::map<Addr, RstEntry>::iterator victim = RST.begin();
+        //Evict region that is dense, if not then LRU
+        std::map<Addr, RstEntry>::iterator candidate;
+        //LRU Victim
+        for (candidate = RST.begin(); candidate != RST.end(); candidate++) {
+           if (candidate->second.lastTouch
+            < victim->second.lastTouch)
+               victim = candidate;
+           if (candidate->second.trained) {
+               victim = candidate;
+               ipcpStats.trainedVictim += 1;
+               break;
+           }
+        }
+
+        if (victim->second.trained) {
+            for (auto &ip : ipcpTable) {
+                if (ip.second.regionAddr == victim->first) {
+                    ip.second.streamValid = true;
+                    ip.second.dir = (victim->second.dir > 32) ? 1 : -1;
+                }
+            }
+        }
+
+        RST.erase(victim);
+    }
+
+   //CS, CPLX
+
+    int64_t curStride = 0;
+
+    //Lookup IPCP
+    if (ipcpTable.find(ipIndex) != ipcpTable.end()) {
+        if (ipcpTable[ipIndex].ipTag == ipTag) {
+            //Hit in IPCP table
+            ipcpTable[ipIndex].valid = 1;
+        }
+        else {
+            //Check valid for taking replacement decision
+            if (ipcpTable[ipIndex].valid == 0) {
+                ipcpTable[ipIndex].ipTag = ipTag;
+                ipcpTable[ipIndex].lastVpage = curPage;
+                ipcpTable[ipIndex].lastLineOffset = lineOffset;
+                ipcpTable[ipIndex].signature = 0;
+                ipcpTable[ipIndex].stride = 0;
+                ipcpTable[ipIndex].conf = 0;
+                ipcpTable[ipIndex].valid = 1;
+                ipcpTable[ipIndex].regionAddr = regionBase;
+                ipcpTable[ipIndex].dir = 1;
+                ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+                 || RST[regionBase].trained);
```

Comment:
This seems to be resetting (or initializing) the entry. This should be a method on the entry since it's used in many different places.